hs_core/models.py

```diff
@@ -3218,25 +3309,27 @@ def url(self):
                 views.file_download_url_mapper,
                 name='get_resource_file')
 
+        This url does NOT depend upon federation status.
         """
-        # must start with a / in order to concat with current_site_url.
-        return '/' + os.path.join('resource', self.resource.short_id,
-                                  'data', 'contents', self.short_path)
+        return '/' + os.path.join('resource', self.public_path)
 
     @property
-    def irods_url(self):
-        """Return the iRODS URL of the file.
+    def public_path(self):
+        """ return the public path (unqualified iRODS path) for a resource.
+            This corresponds to the iRODS path if the resource isn't federated.
+        """
+        return os.path.join(self.resource.short_id, 'data', 'contents', self.short_path)
 
-        This is a direct link and independent of the Django path in ResourceFile.url
-        However, this does not invoke Nginx large file support with sendfile.
-        So use this with caution.
+    @property
+    def irods_path(self):
+        """ Return the irods path for accessing a file, including possible federation information.
+            This consists of the resource id, /data/contents/, and the file path.
         """
-        if self.resource_file:
-            return self.resource_file.url
-        elif self.fed_resource_file:
-            return self.fed_resource_file.url
+
+        if self.resource.is_federated:
+            return os.path.join(self.resource.resource_federation_path, self.public_path)
         else:
-            return None
+            return self.public_path
```

Comment:
I see get_irods_path() method around line 1882 in hs_core/models.py, can they be consolidated in some way?

Reply:
They're different. Public path doesn't have federation information. irods path does.

Reply:
@hyi: 
* public_path is always the unqualified path, i.e., `{resource-id}/data/contents/....`. This returns the path that would be utilized in a URL. 
* get_irods_path() is the qualified path, including federation. It looks like `/{zone}/home/{user]/{resource-id/data/contents/...`
* Because the local files have a default iRODS path, their `public_path` = their `get_irods_path()`. This is not true for federated resources. 

So, `public_path` and `get_irods_path()` have different purposes: 
* `public_path` is for constructing URLs. 
* `get_irods_path()` is for constructing iCommands.

Reply:
@alvacouch OK.  Got it. Thanks for the details.